apps/prairielearn/src/models/enrollment.ts

```diff
@@ -51,22 +76,57 @@ export async function enrollInvitedUserInCourseInstance({
 export async function ensureEnrollment({
   course_instance_id,
   user_id,
+  agent_user_id,
+  agent_authn_user_id,
+  action_detail,
 }: {
   course_instance_id: string;
   user_id: string;
-}): Promise<void> {
-  const user = await selectUserById(user_id);
-  const enrollment = await getEnrollmentForUserInCourseInstanceByPendingUid({
-    course_instance_id,
-    pending_uid: user.uid,
-  });
+  agent_user_id: string | null;
+  agent_authn_user_id: string | null;
+  action_detail: SupportedActionsForTable<'enrollments'>;
+}): Promise<Enrollment | null> {
+  const result = await runInTransactionAsync(async () => {
+    const user = await selectUserById(user_id);
+    const enrollment = await selectOptionalCourseInstanceEnrollmentByPendingUid({
+      course_instance_id,
+      pending_uid: user.uid,
+    });
 
-  if (enrollment && enrollment.status === 'invited') {
-    await enrollInvitedUserInCourseInstance({ course_instance_id, user_id, pending_uid: user.uid });
-    return;
-  }
+    if (enrollment) {
+      await lockEnrollment({ id: enrollment.id });
```

Comment:
PrairieTest also locks users. We should likely do the same. If you don't think it's necessary, please state why.

Nit: PrairieTest has made a habit of passing around complete objects; in this case, that'd be passing all of `enrollment` instead of just the ID. I tend to think this is a good idea.

Reply:
I made the locking users change in 600a09edb16285735d24697687e3e2f2dafa414b.